turn functionalization on in aot_autograd inference

bdhirsh opened the pull request:
still waiting for CI fallout
fixes #90759

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93994
* __->__ #92857



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

bdhirsh committed:
turn functionalization on in aot_autograd inference

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"




[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1172,7 +1228,8 @@ def inp_callable():
 
-        fw_graph = self.verify_aot_autograd(f, inp_callable, test_mutation=True)
+        self.verify_aot_autograd(f, partial(inp_callable, req_grad=False), test_mutation=True)
+        fw_graph = self.verify_aot_autograd(f, partial(inp_callable, req_grad=True), test_mutation=True)
```
it kind of feels like maybe we should have made these tests parametrized

bdhirsh replied:
I started to and ended up not. Partially because `verify_aot_autograd` is already annoyingly complicated, and also because a subset of tests are specifically meant to operate on inputs with mixed requires_gradness. But if you'd prefer the refactor, happy to do it!

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -522,6 +522,18 @@ def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires
+    if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
+        return t
+    torch._sync(t)
+    return torch._from_functional_tensor(t)
```
any modifications?

bdhirsh replied:
none

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1050,9 +1050,52 @@ class AOTConfig:
     num_params_buffers: int
     aot_id: int
 
+def aot_dispatch_base_traced_fn(flat_fn, meta, synthetic_base_info, *flat_args_with_views_handled):
```
This looks suspiciously duped and not deduplicated from the other code

bdhirsh replied:
forgot to leave a comment but yes: the function that we want to trace for `aot_dispatch_autograd` is a superset of what we want to trace for `aot_dispatch_base`, so we can definitely share code.

Originally I felt that `create_joint_forward_backward_functionalized` was already complex enough and trying to factor it to handle the non-autograd case would make it unwieldy, but I think I've factored it in a way that makes it clearer (fresh commit coming in a bit). But lmk what you think

ezyang replied:
To be clear, the original `create_joint_forward_backward_functionalized` is complicated. But I have been pushing for you to figure out how to factor it into smaller composable pieces, to make it less complicated. From adding inference support here, it seems clear that there is a case for making this composable is useful in any case.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1050,9 +1050,52 @@ class AOTConfig:
+    )
+    _input_info = meta.input_info
+
+    flat_args_with_views_handled, _synthetic_base_info = merge_view_inputs(
```
Remind me why we have to merge view inputs for the non-autograd case?

bdhirsh replied:
yep - if you have this code:
```
def f(x, y):
    x.add_(1)
    return x + y

inp = torch.ones(...)
f(inp, inp.view(-1)
```

Then in order for functionalization to pick up the fact that x and y are aliased (and thus the mutation on x should also affect y), we need the graph that we trace to take a synthetic base as an input, that functionalization then sees us take views off of to create the actual inputs.

I guess the one caveat is that the "synthetic base" logic is a lot simpler in the non-autograd case, because we don't have to worry about handling non-differentiable views of differentiable bases. But that just means that all of the cases for synthetic bases should be handled today, in the non-autograd case.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1050,9 +1050,52 @@ class AOTConfig:
+    trace_fn = partial(aot_dispatch_base_traced_fn, flat_fn, meta, _synthetic_base_info)
+    with enable_python_dispatcher():
+        fw_module = make_fx(trace_fn, aot_config.decompositions)(*flat_args_with_views_handled)
+
```
assert no mutable operations?

bdhirsh replied:
yes - thanks

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1065,7 +1108,41 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+        assert len(all_outs) == _num_mutated_inputs + _num_outputs
+
+        if _num_mutated_inputs > 0:
+            nonlocal meta
```
nonlocal? 🤔

bdhirsh replied:
fixed 😛

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1065,7 +1108,41 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+            nonlocal meta
+            assert len(meta.mutated_inp_indices) == _num_mutated_inputs
+            updated_inputs = all_outs[: _num_mutated_inputs]
+            fw_outs = all_outs[_num_mutated_inputs :]
```
Python list slicing is an O(n) operation, best to avoid it in the hotpath

bdhirsh replied:
thanks for the callout - fixed in the updated commit (I have the traced function return a tuple of (updated_inps, fw_outs) so I don't need to bother indexing)

bdhirsh replied:
bleh - I can't easily do this, because inductor codegen expects that the output will be a flat list of tensors

bdhirsh replied:
I'm going to just get everything working first, and come back to this / file an issue

bdhirsh replied:
probably easiest just to manually iterate through the list without doing any slicing (instead of making inductor deal with tuple of list outputs)

ezyang reviewed (approved):
This seems basically reasonable, but I am leary about the code duplication (since I still think there is a more factored implementation that should work) and the hot path execution code feels like it could be tighter. Please check inference perf numbers before and after this patch.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -728,6 +740,24 @@ def inner(*flat_args):
     return inner
 
 
+def unpack_synthetic_bases(primals: List[Any], synthetic_base_info) -> List[Any]:
```
yep I should have done that - will add (and I think I promised to mypy-ify this file a while ago.. maybe I'll come back to that)

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1065,7 +1108,41 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+                    # and use that to mutate the metadata of the input
+                    original_inpt.as_strided_(
+                        updated_inpt.size(),
+                        updated_inpt.stride(),
```
right now, no. I need to think of a good way to do this. We don't actually run the forward pre-functionalization anywhere in AOTAutograd (to use as a reference).

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout




[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout




[ghstack-poisoned]

bdhirsh commented:
I think i've addressed all of the feedback except https://github.com/pytorch/pytorch/pull/92857/files#r1084714158. The important bit is that I updated `create_joint_forward_backward_functionalized` into `create_forward_or_joint_functionalized`, which tries to abstract away the differences between dispatch_base vs dispatch_autograd, and returns the final function that we call make_fx on.

ngimel commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1062,7 +1144,37 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+        assert len(updated_inputs) == _num_mutated_inputs
+        assert len(fw_outs) == _num_outputs
+
+        if _num_mutated_inputs > 0:
```
this is pretty close to what `CompiledFunction.apply` wrapper is doing, are there some helper functions that can be refactored out?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -725,17 +737,42 @@ def inner(*flat_args):
 # (1) Which inputs received data mutations (and need to be passed as outputs into autograd.grad())
 # (2) Which outputs are aliases of inputs (and should *not* be passed as outputs into autograd.grad())
-def create_joint_forward_backward_functionalized(
+def create_forward_or_joint_functionalized(
```
Making a function that previously did one thing do two things toggled by a bool is generally not the way to get clear, understandable code 😛

ezyang replied:
I have audited the `trace_joint` sites and this seems acceptable

bdhirsh replied:
ha - I think that despite the bool, the duplication is pretty minimal. But I could just re-write this as two separate functions calling into the shared logic

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -908,34 +930,60 @@ def joint_forward_backward(
+    def traced_joint(primals, tangents):
+        return functionalized_f_helper((primals, tangents))
+
+    def traced_forward(*primals):
```
Wouldn't it be more symmetric to just have this be `primals` (no asterisk?)

bdhirsh replied:
Yep this took me a bit to untangle, I should add a comment.

The main reason it's currently set up this way is because of the calling convention by the partition. The partition takes in a graph that takes in two arguments `(primals, tangents)`, and emits two graphs, that take in the **unpacked** forward args (and backward args, respectively). So in order to keep the `runtime_wrapper` logic consistent across the autograd case and the inference case, I thought it would be simpler if the final compiled function that we call in both cases has the same convention: it takes in the single arg corresponding to "list of input args", and the calls a compiled fn that expects the unpacked args.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -908,34 +930,60 @@ def joint_forward_backward(
+        outs = flat_fn_with_synthetic_bases_expanded(primals, primals_after_cloning, maybe_tangents)
+        return outs
 
+    def functionalized_f_helper(primals_and_maybe_tangents):
```
and then this can be

```
    def functionalized_f_helper(primals, tangents=None):
```

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -785,46 +822,36 @@ def maybe_to_fresh_input(idx, t):
-                )
-                f_args_inner.append(view_arg)
-        return f_args_inner
+    def forward_or_joint(
```
If possible, I suggest hoisting these helper functions to the top level, as it will make it clearer what variables they closed over versus not, and will make them more easily reusable if they are needed in other contexts.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -908,34 +930,60 @@ def joint_forward_backward(
-    def to_fun(t):
-        if isinstance(t, Tensor):
-            return torch._to_functional_tensor(t, mirror_autograd_meta=True)
+    def flat_fn_with_synthetic_bases_expanded(
```
You've decomposed what previously was one function into three functions. That's a good start... however, you fail to say what the preconditions/postconditions of the newly factored functions are. Without that essential ingredient, factoring the functions doesn't make things easier to understand. For all we know, it may be necessary to read all three functions simultaneously to understand what's going on; and if that's the case, then it would have been better to have one mega function. The pre/post conditions are what enable modular reasoning!

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -785,46 +822,36 @@ def maybe_to_fresh_input(idx, t):
+        mutated_inputs_to_return = [
+            x
+            for (i, x) in enumerate(primals_after_cloning)
+            if meta.input_info[i].mutates_data or meta.input_info[i].mutates_metadata
```
any changes here?

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




[ghstack-poisoned]

bdhirsh commented:
From some of the failing tests: I thought that the inference case wouldn't require special handling for outputs that are aliases of inputs, but that actually isn't quite right. Given this example:
```
def f(x):
    x_view = x.view(-1)
    x_view.add_(1)
    return x_view
```

Functionalization will transform it into:
```
def f(x):
    x_view = x.view(-1)
    x_view_updated = x_view.add(1)
    x_updated = x_view_updated.view(x.shape)
    ...
    x.copy_(x_updated)
    # this is **wrong**: we actually need to regenerate the view off of x and return that
    return x_view_updated
```

So in the case where we have an output that aliases an input, **and** that input was mutated, we need to regenerate the output still. This is slightly different from the autograd, case, where we **always** need to regenerate the output.

ngimel commented:
So this is all looking good, but I wonder what'll happen to optimizers once this is turned on. Optimizers are currently doing inplace updates, that inductor handles, and with this change IIUC inductor will be able to generate only out-of-place updates, that will later be fixed by `copy_`s that are outside of compiled graph. To bring optimizers to their todays state we'd need #91310 implemented too?

eellison commented:
Can we get benchmarks on this ? particularly with the optimizer interaction.

bdhirsh commented:
> To bring optimizers to their todays state we'd need https://github.com/pytorch/pytorch/issues/91310 implemented too?

yep, agreed with all the above. I actually tried kicking off a benchmark yesterday but realized I needed to hose down some failures first. Once I get failures fixed I'll kick off an inference benchmark locally.

Realistically, do we expect that https://github.com/pytorch/pytorch/issues/91310 will need to be implemented as part of this change in order to avoid heavy perf regressions? I could definitely the answering being yes, just don't know for sure.

ngimel commented:
I think the answer is yes, or we need a way to disable functionalization on optimizers to make those changes separable.

bdhirsh commented:
After talking to @ngimel , tentative plan (lmk if anything has thoughts):

(1) benchmark, confirm that we see noticeable regressions on some models due to optimizer step getting slower

(2) the fix: what inductor (most likely) wants is a way to tell aot autograd "please functionalize all internal mutations, but DONT functionalize input mutations". It sounds like inductor would (in general) prefer a functional graph, but for perf reasons, inductor is ok with handling mutations that are known to be input mutations (even more complicated cases, like mutating a view of an input, or handling aliased inputs that are mutated).

Functionalization doesn't actually have the capability to do that today, but adding that functionality should be doable. One way to do it is to store a bit on `FunctionalTensorWrapper` is an "input". We can then add a mode to functionalization (similar to `remove="mutations_and_views"``), that instructs functionalization only to functionalize mutations on tensors that do **not** have that bit set.

One thing that's a bit unclear: we won't be able to run our vanilla DCE anymore in the inference path if we do that. I guess we'll have to benchmark and see if that causes problems.

Finally, aot autograd should still functionalize everything by default. We can add a flag to `aot_autograd` that inductor can specify, that instructs aot autograd to run the version of functionalization that doesn't remove input mutations.

I probably will not bother trying to get that to work for the non-inference case, because (1) the main purpose of all this is to fix optimizer perf, and (2) handling the partitioning and DCE in that case is harder

cc @ezyang for thoughts

ezyang commented:
This plan SGTM

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1626,10 +1975,22 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig):
 
-    joint_forward_backward = create_joint_forward_backward_functionalized(
+    # pre-compute, so we can bail out quickly in the hotpath
+    metadata_ = CompiledRuntimeMetadata(
```
I think that metadata organization could use a cleanup PR - currently planning on revisiting that later

bdhirsh commented:
This is ready for another round.

**Major changes** are:
- I added a flag to aot_module/function, `keep_inference_input_mutations` (off by default, on for inductor)
- That flags instructs the aot_dispatch_base codepath to append `copy_()` calls into the graph, instead of handling input mutations in an epilogue
- I took Natalia's major piece of feedback and merged the runtime wrapper logic into a a shared function, `create_runtime_wrapper`. Most of the epilogue is identical: the only important difference is that the inference codepath doesn't need to handle the "outputs that alias intermediates" case. There's a flag that you pass to `create_runtime_wrapper` that tells it whether or not to handle this case at runtime.


I do think the metadata story is now a bit hairier. Very clearly, the state of the metadata now when this flag is set is the following:
- `CompiledRuntimeMetadata` **still** faithfully records all metadata about which inputs received data mutations / metadata mutations
- However, the bit of metadata that we precompute for runtime, `mutated_inp_indices`, now holds the invariant that it **only** holds the indices of inputs that we need to mutate in the opaque epilogue. So when `keep_inference_input_mutations` is set, this list **only** contains the indices of inputs that need metadata mutations. Normally, it holds the indices of inputs that need either metadata mutations or input mutations

**State of CI**:

I confirmed locally that all `test/functorch/test_aotdispatch.py` fully passes (this is promises, since it contains a ton of weird edge cases).

I tried running `test/inductor/test_torchinductor_opinfo.py` locally, and it segfaults on a triton kernel. Haven't done to debug much yet - I'm going to try cleaning my triton cache and doing a fresh rebuild. In the meantime, CI seems to not be red yet (I'll look at errors as they come up)

I just kicked off a performance run on this PR.

bdhirsh commented on test/inductor/test_torchinductor_opinfo.py:
```diff
@@ -420,6 +416,9 @@ class TestInductorOpInfo(TestCase):
     check_model_cuda = check_model_cuda
 
+    def setUp(self):
+        print("In method", self._testMethodName, flush=True)
```
whoops, will delete before landing

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1033,6 +1138,16 @@ def call_func_with_args(f, args, steal_args=False, disable_amp=False):
+# more API surface for now if this will only be used by AOT Autograd
+from torch.utils._contextlib import _DecoratorContextManager
+
+class KeepInputMutations(_DecoratorContextManager):
```
going to delete this too

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -908,34 +958,89 @@ def joint_forward_backward(
+            #     just see the copy_() directly on x!
+            #     For now, to do this, I'm manually unpacking the synthetic bases again in this function.
+            primals_unpacked = unpack_synthetic_bases(primals, synthetic_base_info)
+            f_primals_unpacked = unpack_synthetic_bases(f_primals, synthetic_base_info)
```
See this comment for details (cc @wconstab since this is related to what you referenced in the meeting).

If you try to compile a kernel like this:
```
def f(x, y):
    x.mul_(2)
    y.mul_(3)
    return x, y

a = torch.ones(1'000'000)
x, y = out(a[0:9], a[1:10])
```

Then:
(1) AOTAutograd will do all of the fancy logic of creating a graph that takes a single argument, corresponding to a "synthetic base" of the aliased x and y arguments
(2) The logic here is what causes us to emit two `copy_()`'s into the graph, for each tiny slice, instead of one massive copy for the (in this case much larger) base.

I think that there are arguments in either direction: in this example, putting the copy_() nodes on the smaller views is more performant, but it means inductor needs to worry about the ordering of the copy_() calls. I'm mainly doing this because it is what AOTAutograd already does today, just in the opaque epilogue.

We could switch things up and make the copies happen on the larger tensor, it would just require some more work - we'd need to actually generate  the "updated synthetic base" that we copy from, which doesn't currently happen.

wconstab replied:
i am not sure its worth doing that. i just wanted to clarify the behavior.  maybe it is not hard for inductor to deal with the copy calls.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
I fixed a few unit tests, although I think... `python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCPU.test_comprehensive_nn_functional_cosine_embedding_loss_cpu_bool` seems to be failing nondeterministically for me locally. I'll see what CI says, but I'm going to nuke my build / triton cache and re-run tomorrow.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
Ok, that test is deterministically failing. I also confirmed that it only fails on inductor, and not aot_eager, and it fails for both cpu and cuda. I have a separate repro here (need to check out my PR first and then run the repro).

Unfortunately I tried the minifier and it didn't work;  it just gave me `RuntimeError: Input graph did not fail the tester`.

Repro code:
```
import torch
from torch import device, dtype

def forward(arg0_1, arg1_1, arg2_1):
    mul = torch.ops.aten.mul.Tensor(arg0_1, arg1_1)
    sum_1 = torch.ops.aten.sum.dim_IntList(mul, [1])
    mul_1 = torch.ops.aten.mul.Tensor(arg0_1, arg0_1)
    sum_2 = torch.ops.aten.sum.dim_IntList(mul_1, [1])
    add = torch.ops.aten.add.Scalar(sum_2, 9.999999960041972e-13)
    mul_2 = torch.ops.aten.mul.Tensor(arg1_1, arg1_1)
    sum_3 = torch.ops.aten.sum.dim_IntList(mul_2, [1])
    add_1 = torch.ops.aten.add.Scalar(sum_3, 9.999999960041972e-13)
    mul_3 = torch.ops.aten.mul.Tensor(add, add_1)
    sqrt = torch.ops.aten.sqrt.default(mul_3)
    div = torch.ops.aten.div.Tensor(sum_1, sqrt)
    full_like = torch.ops.aten.full_like.default(div, 0, dtype = torch.float32, layout = torch.strided, device = device(type='cuda'), pin_memory = False, memory_format = torch.contiguous_format)
    empty_like = torch.ops.aten.empty_like.default(div, memory_format = torch.preserve_format)
    full_like_1 = torch.ops.aten.full_like.default(empty_like, 1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda'), pin_memory = False)
    sub = torch.ops.aten.sub.Tensor(full_like_1, div)
    sub_1 = torch.ops.aten.sub.Scalar(div, -0.09026303813357939)
    maximum = torch.ops.aten.maximum.default(sub_1, torch.tensor(0))
    #_tensor_constant0 = self._tensor_constant0
    #lift_fresh_copy = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0)
    #maximum = torch.ops.aten.maximum.default(sub_1, lift_fresh_copy)
    eq = torch.ops.aten.eq.Scalar(arg2_1, 1)
    where = torch.ops.aten.where.self(eq, sub, full_like)
    eq_1 = torch.ops.aten.eq.Scalar(arg2_1, -1)
    where_1 = torch.ops.aten.where.self(eq_1, maximum, full_like)
    add_2 = torch.ops.aten.add.Tensor(where, where_1)
    return (add_2,)

# Passes with aot_eager
@torch.compile(backend='aot_eager')
#@torch.compile(backend='inductor')
def f1(a, b, c):
    return forward(a, b, c)

a = torch.ones(5, 5, dtype=torch.bool, device='cuda')
b = torch.ones(5, 5, dtype=torch.bool, device='cuda')
c = torch.ones(5, dtype=torch.bool, device='cuda')

out1 = f1(a, b, c)
out2 = forward(a, b, c)

assert isinstance(out1, tuple)
assert isinstance(out2, tuple)
assert len(out1) == len(out2)
for o1, o2 in zip(out1, out2):
    assert torch.allclose(o1, o2)
```

Even worse... I can't get the segfault to repro from a single test. Instead, I see a segfault consistently when I run multiple tests. I also checked gdb, and the segfault is deep in triton somewhere I tried clearing my triton cache with  `rm -rf /tmp/torchinductor_hirsheybar/`.

I also tried nuking and re-installing my triton, based on the current commit hash in `triton.txt`:
```
pip install "git+https://github.com/openai/triton@0d7e7532279e45672555e344646f5c19c3972331#subdirectory=python"
```

@ngimel do you think you'd be able to take a look? Or if you have some ideas for other things I can try, happy to try them.

Segfault repro command is:
```
python test/inductor/test_torchinductor_opinfo.py -k TestInductorOpInfoCUDA.test_comprehensive_nn_functional_cosine_embedding_loss_cuda
```

bdhirsh commented:
There are still more failures from CI to hit, so I'll start to look over them.

bdhirsh commented:
marked a few failures as low-pri for now. Currently looking at the tts_angular failure.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -436,11 +436,21 @@ class ViewAndMutationMeta:
     num_intermediate_bases: int
 
+    # For inference only: instructs us to keep data-only input mutations directly in the graph
+    keep_input_mutations: int
```
Int?

ezyang commented on torch/_inductor/decomposition.py:
```diff
@@ -392,6 +386,12 @@ def bernoulli_(self, p=0.5):
     return self.copy_(torch.rand_like(self, dtype=torch.float32) < p)
 
 
+@register_extra_random_decomp([aten.bernoulli.p])
```
?

bdhirsh replied:
Hmmm let me try to find the original test failure that caused me to add this...

Context: I had a bernoulli test failing on inductor somewhere, and I noticed that inductor already does something similar for the inplace variant of bernoulli ([code](https://github.com/pytorch/pytorch/blob/fe0e28ab87ddda38a82efe74dd3eaf05e926f117/torch/_inductor/decomposition.py#L386)). I thought it would make sense to copy the pattern for the out-of-place variant, since it looks like the purpose of this decorator is precisely to handle the fact that randomness-related decomps can cause divergences from eager behavior.

bdhirsh replied:
It was this test: `python test/inductor/test_torchinductor.py CpuTests.test_bernoulli1_cpu`. I'll confirm with @ngimel: does adding this decomp to inductor's "randomness" decomps sound like the right thing here?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -496,7 +506,12 @@ def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires
+        # This shape mismatch can happen due to a bug in inplace/view handling in autograd.
+        # Try putting a breakpoint here and running `test/functorch/test_aotdispatch TestAOTAutograd`
+        # To see where that happens.
+        # Also, https://github.com/pytorch/pytorch/issues/49825
```
Tbh, the issue isn't terribly clear about what the problem is. An update would be helpful. Cc @albanD

bdhirsh replied:
Yep, now that we have the context manager for turning on view replay this is easier. I added a repro in https://github.com/pytorch/pytorch/issues/49825#issuecomment-1422865300.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -496,7 +506,12 @@ def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires
         out = target_meta_tensor._view_func(reshaped_base_tensor)
-        if out is not None:
+        # This shape mismatch can happen due to a bug in inplace/view handling in autograd.
+        # Try putting a breakpoint here and running `test/functorch/test_aotdispatch TestAOTAutograd`
```
Did you mean to quote a specific test?

bdhirsh replied:
yep - added

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -908,34 +958,89 @@ def joint_forward_backward(
+            #     For example, suppose "x" and "y" are aliased inputs, and we witness:
+            #       x.add_(1)
+            #       y.t_()
+            #     We don't actually want the compiler to witness a copy_() on the synthetic base here: it should
```
The copy on synthetic base is akin to directly updating the storage. Maybe it is marginally better to tell the compiler about mutation on x, but updating the storage also seems sound to me?

ezyang replied:
I guess this is what you say in the comment below. Your example on GH is much better than the example in the code here

bdhirsh replied:
Yes agreed, I'll put that directly in the code comment

ezyang reviewed (commented):
It would have been a lot easier to review if you had first landed the version of this PR without input mutation, but without turning on inference functionalization by default, and then a separate PR to add input mutation in graph

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1668,6 +1756,229 @@ def describe_input(i, aot_config):
+# This is because there are some minor differences in how we treat these cases at runtime:
+# - resize_() is currently handled in the inference case, but not fully handled in the autograd case.
+# - the autograd cases inserts TensorAlias wrapper objects for outputs that alias inputs
+def create_runtime_wrapper(
```
What modifications were made here when it was moved?

bdhirsh replied:
No functional changes. The only difference was that I gave it the extra argument, `keep_input_mutations` - and the only reason for that is because the debug asserts change slightly when you use `keep_input_mutations`.

The reason that there are no functional changes in the runtime wrapper, is because the wrapper goes off of `runtime_metadata.fw_metadata.mutated_inp_indices` to know which indices to apply a mutation with. And our trace-time logic has already ensured that data-only mutation indices don't show up in this list when `keep_input_mutations` is set. So the wrapper itself doesn't need to do anything different - it just blindly applies updates to the inputs based on that list of indices.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1808,16 +2093,8 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig):
-        num_outputs_aliased_to_inputs = _num_outputs_aliased_to_inputs
-        num_outputs_aliased_to_intermediates = _num_outputs_aliased_to_intermediates
-        num_outputs_aliased = _num_outputs_aliased
+        metadata = metadata_
```
This refactor could have been done in a prep commit and landed super easily. Folded in here the reviewers have to slog through this along with all of the other changes in the diff.

bdhirsh replied:
Yes. In retrospect, I totally should have done that - sorry about that!

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -970,6 +998,35 @@ def functionalized_f_helper(primals_and_maybe_tangents):
+                    # so the compiler will see the input mutation in the graph.
+                    assert inpt_new is not inpt_old
+                    assert has_same_metadata(inpt_new, inpt_old)
+                    inpt_old.copy_(inpt_new)
```
What I was hoping to see here, code structure wise, was a function that "applies copy_" and it being called from both the runtime wrapper, or here, guarded by the `keep_input_mutation` flag which controlled where you ran it. This would be make it obvious that the same logic was being used in the wrapper as in codegen. On the plus side I suppose this code is not that long. But it is not at all clear from the code itself (without your PR comment in https://github.com/pytorch/pytorch/pull/92857/files/6bdf2563239b88ce9dbbe52afbde3e37a2e24412..c764f7fa0478e24176bc422888a44ae391d1b1fa#r1095151376 which will be lost to the sands of time when this PR lands) that this is what's going on

bdhirsh replied:
> What I was hoping to see here, code structure wise, was a function that "applies copy_" and it being called from both the runtime wrapper, or here

Definitely agreed that a shared function like this, even if it doesn't deduplicate much code, would make things clearer (you can read the code and understand that the copy_() calls happen either at runtime or trace time, gated on a flag).

Definitely open to feedback - but the main reason I didn't end up trying to refactor in that way was perf. Today, at trace time we pre-compute a set of "input indices" that we need to loop over in the runtime epilogue, to handle any data / metadata mutations (`runtime_metadata.fw_metadata.mutated_inp_indices`). I opted to pre-compute this list to **not**  include the data-only mutations when `keep_inference_input_mutations` is set ([code comment](https://github.com/pytorch/pytorch/pull/92857/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R446)). Concretely, if we we compile an optimizer step with 100 data-only input mutations, then this will avoid looping over 100 indices that we no-op on.

Either way, for sure this set of conventions and assumptions in the code is complicated. My thought is that is that we want both better code clarity and we don't want to sacrifice runtime overhead, so I can add some clearer comments (I have some already, that can probably be improved).

ezyang replied:
I buy the precompute argument. Leave it as is.

ezyang replied:
If we really really wanted it to happen, you'd probably have to arrange for the precompute to happen at compile time too, so you can call the precompute copy apply.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1140,49 +1210,24 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
 
-    new_fn._boxed_call = True
+    # TODO: I think that now that the new function takes in unpacked args (compiled_fn(*args)), I don't need this anymore.
+    # need to confirm.
```
TODO

bdhirsh replied:
I looked at the inductor output from compiling optimizers in `test/dynamo/test_optimizers.py` (forcing it to run with inductor), and confirmed that the generate kernels do the list-clearing properly (they take in a single list arg, unpack and clear it).

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
             example_idx = aliased_input_indices[0]
-            synthetic_base = torch.Tensor(fwd_inputs[example_idx].untyped_storage())
+            if is_runtime:
+                # Wondering when we hit this case? a known example is when compiling
```
What is this case?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
+                # We don't actually have a convenient way of going from storage -> tensor,
+                # So using set_() here (we suffer some minor overhead, but this case is rare).
+                synthetic_base = torch.empty((), dtype=example_alias.dtype, device=example_alias.device)
+                synthetic_base.set_(example_alias.untyped_storage())
```
this is a behavior change from before, I guess?

bdhirsh replied:
The "old behavior" was just that this case was broken and would error - "this case" being "inputs share storage, but none of them have a `._base` attribute (it would error because my original code tried to construct `torch.Tensor(storage)` with a cuda storage, which is not allowed).

I spent some time looking at the LSTM code and actually couldn't really find the logic that set things up so that there was no ._base attribute on the aliased tensors. So locally, I have a new test that just simulates this behavior by calling .detach() on the views before passing them in (going finish going through PR feedback before pushing more changes).

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
+                example_alias = fwd_inputs[example_idx]
+                # We don't actually have a convenient way of going from storage -> tensor,
+                # So using set_() here (we suffer some minor overhead, but this case is rare).
+                synthetic_base = torch.empty((), dtype=example_alias.dtype, device=example_alias.device)
```
Don't pass () to empty, that will incur a 1 element allocation. Pass (0,) instead.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
+            if is_runtime:
+                # Wondering when we hit this case? a known example is when compiling
+                # the optimizer step, where the parameters came from an LSTM module.
+                # LSTM performs weight packing, so weights are views from a shared buffer.
```
This isn't a complete explanation. Try something like: "Under what circumstance can we have tensors that share a storage, but they do not have the _base field? The _base field simply says that autograd knows about the aliasing relationship, but sometimes we create tensors which are aliased out of the same storage but guaranteed to be disjoint--in these cases, we will skip setting up the _base relationship for performance reasons (because the fact that the tensors share the same storage is unobservable unless you (1) do naughty things with resize_/as_strided or (2) look at the storage--as we are doing here.) One particular example of this is optimizer steps on the LSTM module: LSTM parameters are packed into a contiguous storage for efficiency reasons when calling cuDNN kernels, so when these parameters get passed to the optimizer we will find they share the same storage, but do not have _base set since they are all disjoint."

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1347,7 +1400,37 @@ def merge_view_inputs(
+                # so just plumb the storage through a temporary real tensor.
+                synthetic_base_dummy = torch.Tensor(fwd_inputs[example_idx].untyped_storage())
+                synthetic_base_ = curr_fake_mode.from_tensor(synthetic_base_dummy)
+                synthetic_base = synthetic_base_.to(device=curr_aliased_inp.device)
```
The way we create a fake tensor with a specific storage in meta_utils.py is by allocating a fake tensor, and then set_'ing the fake storage into it. It's not clear why you can't do that here. In fact, this would permit you to use precisely the code in the `is_runtime` branch.

bdhirsh replied:
Thanks - I played around with doing this for a bit, and realized the situation is even better. When this function is run under tracing, `synthetic_base = torch.empty((0,), dtype=example_alias.dtype, device=example_alias.device)` which just do the right thing already and create a FakeTensor, since our entire trace function is run under FakeTensorMode.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1668,6 +1756,229 @@ def describe_input(i, aot_config):
     else:
         return f"input {i - aot_config.num_params_buffers}"
 
+# TODO: document this better
```
TODO

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1668,6 +1756,229 @@ def describe_input(i, aot_config):
+            [
+                x
+                for x in self.fw_metadata.input_info
+                if not x.mutates_data and x.mutates_metadata
```
This looks like just refactor, no semantics change

bdhirsh replied:
That's correct. I just had existing metadata collected by `run_functionalized_fw_and_collect_metadata` (the `ViewAndMutationMeta`), and then an additional set of metadata around synthetic bases. I need the synthetic base metadata in more places, so I wanted to jam them all together.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1668,6 +1756,229 @@ def describe_input(i, aot_config):
+    def runtime_wrapper(*args):
+        # Step 2: remove aliased inputs that are mutated, replace with synthetic bases
+        # Only happens if our graph mutates an input that aliases another input.
+        if runtime_metadata.synthetic_base_info is not None:
```
cross-ref this with https://github.com/pytorch/pytorch/pull/92857/files/6bdf2563239b88ce9dbbe52afbde3e37a2e24412..c764f7fa0478e24176bc422888a44ae391d1b1fa#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416L1146

bdhirsh replied:
Hmmm not sure I understand? That links to line 1146, which is the `KeepInputMutations` decorator (that I'm going to delete in a bit - that was when I went with the first approach of trying to have functionalization ignore input mutations entirely).

bdhirsh commented:
I'll focus on the existing CI failures first, then pivot to the (very valid) PR feedback. On the errors:

(1) most concerning was the `sebotnet33ts_256` accuracy failure. I have that + a few other tests fixed locally, it was due to functionalization not handling channels last contiguity.

(2) Added a few expected failures for some tests that expect us to raise an error, but functionalization doesn't raise the error properly (e.g. when we convert copy_() to copy()). Should fix, but not hi-pri.

(3) What I'm stuck on right now: a few inductor-specific failures.


The first is these 3 tests all pass individually:
```
python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_nn_functional_softmin_cuda_float16
python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_masked_softmin_cuda_float16
python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_softmax_cuda_float16
```

But the three of them all have accuracy failures when I run them together via `-k softm`. Are there some inductor caches that might cause problems here? Running the exact same set of tests under the `aot_eager` backend runs successfully. Repro command:
```
python test/inductor/test_torchinductor_opinfo.py -k _softm
```

The second is an inductor-only failure that I was able to create a min-repro for. It only fails when the inputs have `torch.bool` dtype, and passes when the inputs have e.g. `dtype=torch.float32`. I checked, and the graph that we send to inductor is identical in both cases.
```
import torch

def f(a, b, c):
    return torch.nn.functional.cosine_embedding_loss(a, b, c, reduction='none')

# We only fail when using inductor + dtype=bool (fail on both cpu and cuda)
# When backend is aot_eager, or any other dtype is used, we pass.
#
device = 'cuda'
backend = 'inductor'
dtype = torch.bool

a = torch.ones(5, 5, device=device, dtype=dtype)
b = torch.ones(5, 5, device=device, dtype=dtype)
c = torch.ones(5, device=device, dtype=dtype)

f_compiled = torch.compile(backend=backend)(f)
out = f(a, b, c)
out_compiled = f_compiled(a, b, c)
print(torch.allclose(out, out_compiled))
```

@ngimel do you happen to have any suggestions for looking into these problems, or bandwidth to take a look? I'm particularly worried about the first one, since it's harder to create a min repro and seems to be a problem with inductor statefullness. I'm going to dig into the `torch.bool` repro first, since it's more self-contained.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1047,9 +1095,43 @@ class AOTConfig:
+    # aot_dispatch_base requires functionalization, but doesn't need to handle as many cases as the autograd case.
+    # The cases that aot_dispatch_base doesn't need to handle include:
+    # - outputs that are aliases of graph intermediates
+    # - outputs that are aliases of graph inputs
```
(answer: they do, 😛 this is fixed)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1047,9 +1095,43 @@ class AOTConfig:
+        trace_joint=False,
+    )
+    with enable_python_dispatcher():
+        # second arg None is what tells our trace_fn to trace only the forward, and not the joint
```
legacy comment, deleting (thanks!
)

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
Updated. I addressed + commented on all of the targeted PR feedback. The majority of the material changes were:
- Moved create_forward_or_joint_functionalized and it’s layered functions all to be global scope (so it’s clearer whether or not they use state from the outer context), and wrote clearer pre-conditions for them
- Updated create_forward_or_joint_functionalized to take in `CompiledRuntimeMetadata`, instead of separetely taking in `ViewAndMutationMeta` + synthetic_base_info
- Fixed the “generate synthetic base when there is no ._base attribute” case so it’s the same in trace time and run time

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
Have you solved the inductor problems?

bdhirsh commented:
After talking to Natalia, the failures that I sent to her all seemed to be either flaky (existing before this PR too), or mild accuracy differences where we're ok upping the tolerance on some tests (specific to float16 tests). For what it's worth, I'm still not entirely clear on what the flakiness story is, where a few tests seem to have always been broken, and haven't broken master CI for some reason. Just to enumerate them:


** existing failures / suspicious flakiness**

(1) rqsrt test with i32 (mac only) - This was failing in CI on mac only, and the b8 version of this test was already suspiciously skipped only for mac, so I added i32 to this list too.

(2) `nn.functional.cosine_embedding_loss` with `torch.bool` dtype - I confirmed with Natalia that this test is able to fail on a commit before my PR. This test seems flaky though - I xfail'd it originally, and got an unexpected test on my latest CI run. I'll mark it as skipped.

(3) `as_strided.partial_views` with a few more dtypes. This test is already xfail'd with a few other dtypes, although I believe that this case is already pretty broken with AOTAutograd (and is difficult to fix). This feels low-pri, since it only matters if a user calls `.as_strided()` in the graph we compile, and sends us inputs with weird striding + non-zero storage offsets.


**suspicious accuracy differences**
`softmax`, `softmin` and `masked.softmin` all have minor accuracy failures, only for `float16`. Natalia agreed we can chalk it up to f16 just legitimately needed a wider testing tolerance, although it's unclear why functionalizing the graphs in my PR supposedly makes this worse.

bdhirsh commented:
`jx_nest_base` also failed on the last CI run when run with dynamic shapes, but passes when I run it locally. The error also seems pretty unrelated to this PR:
```
2023-02-09T21:01:57.4163223Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 224, in wrapped
2023-02-09T21:01:57.4163635Z     return decomp_fn(*args, **kwargs)
2023-02-09T21:01:57.4164235Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/experimental/symbolic_shapes.py", line 497, in <lambda>
2023-02-09T21:01:57.4164630Z     'sym_int': lambda a: error(),
2023-02-09T21:01:57.4165035Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/experimental/symbolic_shapes.py", line 483, in error
2023-02-09T21:01:57.4165396Z     raise AssertionError("shouldn't be hit")
2023-02-09T21:01:57.4165675Z AssertionError: shouldn't be hit
2023-02-09T21:01:57.4179204Z ERROR:common:inductor raised LoweringException: AssertionError: shouldn't be hit
2023-02-09T21:01:57.4179642Z   target: <function sym_int at 0x7f665f61dab0>
2023-02-09T21:01:57.4179940Z   args[0]: 4
```

bdhirsh commented:
@ngimel it looks like `python test/inductor/test_torchinductor.py CudaTests.test_argmax_argmin2_cuda` is flaky too - confirmed locally that it errors for me on the commit below my PR.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on test/inductor/test_torchinductor.py:
```diff
@@ -5017,6 +5017,11 @@ def fn(x):
 
+    @unittest.skip(
+        """
+        FIX: this test is flaky
```
file a bug for it and link?

bdhirsh replied:
I tried to repro again locally in order to include an error in the issue, and realized I can't repro anymore :(. I think I'll remove the skip and see if CI passes for sanity (and if it does, I'll just land this PR - since the flakiness of this test is independent). If it comes up again, I'll skip and file an issue.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -776,225 +777,336 @@ def unpack_synthetic_bases(
+@dataclass
+class CompiledRuntimeMetadata:
+    # This type / object should be cleaned up
+    synthetic_base_info: Optional[List[Union[int, Tuple[int, torch.Tensor]]]]
```
xref this with where the info is documented

bdhirsh replied:
done

voznesenskym commented on torch/_functorch/aot_autograd.py:
```diff
@@ -436,11 +436,21 @@ class ViewAndMutationMeta:
-        mutated_inp_indices = [
-            i for i, m in enumerate(self.input_info) if m.mutates_data or m.mutates_metadata
-        ]
+        if self.keep_input_mutations:
+            # When keep_input_mutations is set, we don't need to worry about our epilogue
+            # handling data-only mutations, because we keep them directly in the graph.
+            mutated_inp_indices = [
+                i for i, m in enumerate(self.input_info) if m.mutates_metadata
+            ]
+        else:
+            mutated_inp_indices = [
+                i for i, m in enumerate(self.input_info) if m.mutates_data or m.mutates_metadata
+            ]
```
Could this just be:

```suggestion
          # When keep_input_mutations is set, we don't need to worry about our epilogue
          # handling data-only mutations, because we keep them directly in the graph.
          mutated_inp_indices = [
              i for i, m in enumerate(self.input_info) if m.mutates_metadata or (self.keep_input_mutations and  m.mutates_data)
          ]
```

bdhirsh replied:
good catch, thanks - (it should actually be `if m.mutates_metadata or (not self.keep_input_mutations and  m.mutates_data)` -  "keep_input_mutations" means "keep input mutations in the *graph*, so we should *not* include them in the list of mutated input indices to handle in the epilogue")

voznesenskym commented on torch/_functorch/aot_autograd.py:
```diff
@@ -719,223 +749,420 @@ def inner(*flat_args):
+#     that has multiple input aliases.
+# (3) If any of those corresponding inputs get metadata mutations, then we clone the base.
+def maybe_to_fresh_input(idx, t, meta):
+    if not isinstance(t, Tensor):
```
When is this not a tensor?

bdhirsh replied:
Ignoring dynamo for a second - if you directly pass a non-tensor input as an example input into `aot_module_simplified` or `aot_function`, it'll get plumbed down to here (AOTAutograd flattens pytrees, but it doesn't do anything to filter out non-tensor inputs). E.g. if you pass an `int` int (test [here](https://github.com/pytorch/pytorch/blob/a27bd42bb9ad39504fdd94ad38a5ad0346f1758b/test/functorch/test_aotdispatch.py#L334)).

Dynamo won't pass in int inputs, but some people are directly using the AOTAutograd API's, so I guess it's safer to be resilient to them.

SymInt is probably a better example (you can reasonably have a graph where its inputs are SymInts? Ed/Horace had some discussion around whether or not to establish some invariants around all SymInt inputs getting promoted to 0-dim tensors, although that's not the case today).

voznesenskym commented on torch/_functorch/aot_autograd.py:
```diff
@@ -719,223 +749,420 @@ def inner(*flat_args):
+        if is_grad_tensor:
+            grad_primals.append(p)
+
+    # Get the outputs that need gradients
+    assert len(tangents) == len(fw_outs_to_grad)
```
nit: can we move this assert up?

bdhirsh replied:
yep, done

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
Interesting new failure that didn't show up before: There's an input mutation test in inductor that involves mutating the input twice, that seems to fail on the CPU backend but on the CUDA (triton) backend. Looking into it more now. repro:
```
python test/inductor/test_torchinductor.py CpuTests.test_slice_mutation2_cpu
```

bdhirsh commented:
I have a repro that doesn't include AOTAutograd - it only includes inductor's `compile_fx_inner`. It also looks like my repro passes on a commit from 2 days ago, `76ed1a81d14f18d6078f11d525aafe5de694cadb`, but it fails on master.

Repro:

```
import torch
from torch._inductor.compile_fx import compile_fx_inner
from torch.fx.experimental.proxy_tensor import make_fx

def f(arg0_1):
    slice_1 = torch.ops.aten.slice.Tensor(arg0_1, 0, 0)
    slice_2 = torch.ops.aten.slice.Tensor(slice_1, 1, 20, 40)
    add = torch.ops.aten.add.Tensor(slice_2, 1)
    slice_3 = torch.ops.aten.slice.Tensor(arg0_1, 0, 0)
    slice_4 = torch.ops.aten.slice.Tensor(slice_3, 1, 20, 40)
    slice_5 = torch.ops.aten.slice.Tensor(arg0_1, 0, 0)
    slice_scatter = torch.ops.aten.slice_scatter.default(slice_5, add, 1, 20, 40)
    slice_scatter_1 = torch.ops.aten.slice_scatter.default(arg0_1, slice_scatter, 0, 0)
    slice_6 = torch.ops.aten.slice.Tensor(slice_scatter_1, 0, 0)
    slice_7 = torch.ops.aten.slice.Tensor(slice_6, 1, 20, 40)
    slice_8 = torch.ops.aten.slice.Tensor(arg0_1, 0, 0)
    slice_9 = torch.ops.aten.slice.Tensor(slice_8, 1, 1, 10)
    slice_10 = torch.ops.aten.slice.Tensor(slice_scatter_1, 0, 0)
    slice_11 = torch.ops.aten.slice.Tensor(slice_10, 1, 1, 10)
    add_1 = torch.ops.aten.add.Tensor(slice_11, 2)
    slice_12 = torch.ops.aten.slice.Tensor(arg0_1, 0, 0)
    slice_13 = torch.ops.aten.slice.Tensor(slice_12, 1, 2, 11)
    slice_14 = torch.ops.aten.slice.Tensor(slice_scatter_1, 0, 0)
    slice_15 = torch.ops.aten.slice.Tensor(slice_14, 1, 2, 11)
    slice_16 = torch.ops.aten.slice.Tensor(slice_scatter_1, 0, 0)
    slice_scatter_2 = torch.ops.aten.slice_scatter.default(slice_16, add_1, 1, 2, 11)
    slice_scatter_3 = torch.ops.aten.slice_scatter.default(slice_scatter_1, slice_scatter_2, 0, 0)
    slice_17 = torch.ops.aten.slice.Tensor(slice_scatter_3, 0, 0)
    slice_18 = torch.ops.aten.slice.Tensor(slice_17, 1, 2, 11)
    copy_ = torch.ops.aten.copy_.default(arg0_1, slice_scatter_3)
    return ()

x_ref = torch.ones([1, 64], device='cpu')
x_test = torch.ones([1, 64], device='cpu')
x_test2 = torch.ones([1, 64], device='cpu')
x_test3 = torch.ones([1, 64], device='cpu')


fx_g = make_fx(f)(x_test2)
f_compiled = compile_fx_inner(fx_g, [x_test3])

f(x_ref)
f_compiled([x_test])

print(x_test)
print(torch.abs(x_test - x_ref))
# prints False
print(torch.allclose(x_test, x_ref))
```

bdhirsh commented:
It looks like https://github.com/pytorch/pytorch/pull/94110 is the culprit - the above repro passes on the commit before on master (`7ce785b50b15e50b5aff9f62451a0c1f01b03f03`), but fails on that commit (ceab30775b80306f10a08b2f1e3a4d11b1835a75). Filing an issue.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
The other failing set of tests was `python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCPU.test_comprehensive_resize__cpu_float16` - I just added an explicit fallback for `resize` and `resize_as` to `lowerings.py` to get around it (I think there was a recent inductor PR that stopped us from adding implicit fallbacks for missing ops).

I also xfail'd the test above for now just to confirm that I can get green CI, but I'll wait for some discussion before landing anything.

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "turn functionalization on in aot_autograd inference"

still waiting for CI fallout
fixes #90759




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

seemethere commented on torch/_functorch/aot_autograd.py:
```diff
@@ -436,10 +436,15 @@ class ViewAndMutationMeta:
     num_intermediate_bases: int
 
+    # For inference only: instructs us to keep data-only input mutations directly in the graph
+    keep_input_mutations: int
```
It would be better to have a default value here instead of making a backwards incompatible change.

```python
    keep_input_mutations: int = 0
```

seemethere commented on torch/_functorch/aot_autograd.py:
```diff
@@ -541,6 +564,8 @@ def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires
 def run_functionalized_fw_and_collect_metadata(
     f,
+    *,
+    keep_input_mutations: bool
```
```python
    keep_input_mutations: bool = False
```

seemethere commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1046,11 +1268,50 @@ class AOTConfig:
     num_params_buffers: int
     aot_id: int
-
+    keep_inference_input_mutations: bool
```
```python
    keep_inference_input_mutations: bool = False
```

seemethere reviewed (commented):
We should probably revert this and re-submit with sane defaults for all of the new keyword arguments since it's failing internal tests. In the future it'd be good to be mindful of adding new required keyword arguments since they can break compatibility.

weiwangmeta commented:
@pytorchbot revert -m "breaking internal builds due to API backward incompatibility" -c ghfirst

ezyang commented:
@seemethere these are supposed to be private data structures lol
